src/dotty/tools/dotc/core/TypeApplications.scala

```diff
@@ -524,21 +524,28 @@ class TypeApplications(val self: Type) extends AnyVal {
     }
   }
 
-  /** Convert a type constructor `TC` with type parameters `T1, ..., Tn` to
+  /** Convert a type constructor `TC` which has type parameters `T1, ..., Tn`
+   *  in a context where type parameters `U1,...,Un` are expected to
    *
    *     LambdaXYZ { Apply = TC[hk$0, ..., hk$n] }
    *
-   *  where XYZ is a corresponds to the variances of the type parameters.
+   *  Here, XYZ corresponds to the variances of
+   *   - `U1,...,Un` if the variances of `T1,...,Tn` are pairwise compatible with `U1,...,Un`,
+   *   - `T1,...,Tn` otherwise.
```

Comment:
I think it can also appear in tryLifted, which gets called during a subtype check. Essentially, we compare an unexpanded hk type with a type lambda and go on expanding the unexpanded type. It seems we can get a discrepancy in variances there.